### PR TITLE
Add ISchema.Comparer

### LINKED ISSUE #1905: Field ordering in introspection queries
Hi, I see that as of #594 the fields of a type in a schema are always ordered alphabetically. This is somewhat confusing with a schema-first approach to a server as the fields are all out of order from their originally shared schema. Could we please make this behaviour optional? It would make sense to me to control ordering by providing the user a setter for an `IComparer<FieldType>` which defaults to the original ordering and having a "ByName" implementation provided OOTB to smooth the common case.

Or at the very least can the behaviour be opt-out if not opt-in?

### PR DESCRIPTION
**Pull request by sungam3r**
fixes #1905 

Also adds a bunch of xml comments.

### CODE REVIEW

**Comment by Shane32** [comment]
I like the PR....except I don't understand how this fulfills #1905 - there is nothing provided to preserve ordering for schema-first schemas.  So it doesn't seem to have any practical usefulness as of yet.

Before merging this, I think we should finish the code required to preserve ordering for schema-first schemas, and then tests should be added.  I would not merge it quite yet, unless you plan to follow up with another PR very shortly.

**Comment by sungam3r** [comment]
```c#
 public class MySchemaBuilder : SchemaBuilder
    {
        protected override void PreConfigure(Schema schema)
        {
            base.PreConfigure(schema);
            schema.Comparer = new MyComparer();
        }
    }
...
 var b = new MySchemaBuilder();
 b.Build("");
```

**Comment by sungam3r** [comment]
`Schema.For` is just a shortcut for the default builder
```c#
  public static ISchema For(string typeDefinitions, Action<SchemaBuilder> configure = null)
        {
            var builder = new SchemaBuilder();
            configure?.Invoke(builder);
            return builder.Build(typeDefinitions);
        }
```

**Comment by Shane32** [comment]
If a default comparer (with all null properties) will preserve ordering, then we should write a test case for that.

**Comment by sungam3r** [comment]
Not all null, field comparer is not null and compares fields by their names. I will add tests later.

**Review by joemcbride — APPROVED** [review]
Additional "default" test would be good, but looks good.

A thought.  Should this also be used in the `SchemaPrinter`?  🤔

**Comment by sungam3r** [comment]
The same concept _may be used_ in `SchemaPrinter`.

File: docs2/site/docs/guides/migration4.md
Comment (Shane32) [inline docs2/site/docs/guides/migration4.md:9]:
```suggestion
* `GraphQL.Utilities.ServiceProviderExtensions` has been made internal. This affects usages of its extension method `GetRequiredService`. Instead, reference the `Microsoft.Extensions.DependencyInjection.Abstractions` NuGet package and use the extension method provided by the `Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions` class.
```

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (sungam3r) [inline src/GraphQL/Introspection/ISchemaComparer.cs:43]: @Shane32 we have already discussed this. I am in favor of removing the default sorting in v4. `LegacyV3SchemaComparer` can be used to switch to the old behavior.

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (sungam3r) [inline src/GraphQL/Introspection/ISchemaComparer.cs:70]: Do we need this class?

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (sungam3r) [inline src/GraphQL/Introspection/ISchemaComparer.cs:86]: I'm not sure about naming - _Alphabetically_.

**Comment by sungam3r** [comment]
I added new label for PRs with new APIs with non breaking changes.

**Comment by sungam3r** [comment]
Both new alerts from lgtm bot are false positives.

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (Shane32) [reply src/GraphQL/Introspection/ISchemaComparer.cs:48]: Sounds fine to me.

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (Shane32) [reply src/GraphQL/Introspection/ISchemaComparer.cs:70]: Can we just include the source code for this class in the migration notes in case someone needs it?

Comment (sungam3r) [reply src/GraphQL/Introspection/ISchemaComparer.cs:70]: done

**Review by Shane32 — APPROVED** [review]

File: src/GraphQL/Introspection/ISchemaComparer.cs
Comment (Shane32) [reply src/GraphQL/Introspection/ISchemaComparer.cs:69]: I agree.  Don't have any better suggestions.  Perhaps `AlphabeticalSchemaComparer` ?

Comment (sungam3r) [reply src/GraphQL/Introspection/ISchemaComparer.cs]: Done

**Pull request merged**
